core/plugin_api.rs

```diff
@@ -11,13 +11,37 @@
 pub use crate::Buf;
 pub use crate::Op;
 pub use crate::OpId;
+pub use crate::Resource;
+pub use crate::ResourceId;
 pub use crate::ZeroCopyBuf;
 
+use std::cell::Ref;
+use std::cell::RefMut;
+
 pub type InitFn = fn(&mut dyn Interface);
 
 pub type DispatchOpFn =
   fn(&mut dyn Interface, &[u8], Option<ZeroCopyBuf>) -> Op;
 
+/// Equivalent to ResourceTable for normal ops, but uses dynamic dispatch
+/// rather than type parameters for the `get`, `get_mut`, and `remove` methods.
+pub trait ResourceTable {
+  fn add(&mut self, name: &str, resource: Box<dyn Resource>) -> ResourceId;
+  fn get(&self, rid: ResourceId) -> Option<Ref<dyn Resource>>;
```

Comment:
What I like better about @afinch7's original patch is that `get()` (well, `get_boxed()` that is) returns an `Option<&dyn Resource>`.

Unfortunately the trade off there is that `resource_table()` returns some boxed contraption. Ideally `resource_table()` would return a `RefMut<dyn plugin_api::ResourceTable>` but that probably means the implementation of `plugin_api::ResourceTable for core::ResourceTable` needs to move to core; rust doesn't permit the impl trait to sit in cli because neither the trait nor the type it is implemented for are defined in cli (they're both in core).

Reply:
Fixed. Best of both worlds now.